File: EmployeeManagement_System/wwwroot/databaseService.js
class DatabaseService {
    constructor() {
        this.baseUrl = '/api'; // This would be your backend API endpoint
    }

    // Simulated API calls - in a real app, these would be fetch/axios calls to your C# backend
    async Signup(username, password, email) {
        // Simulate API call delay
        await this.simulateDelay();

        // In a real implementation, this would be:
        // const response = await fetch('/api/auth/signup', {
        //     method: 'POST',
        //     headers: { 'Content-Type': 'application/json' },
        //     body: JSON.stringify({ username, password, email })
        // });
        // return await response.text();

        // Simulated response
        const users = this.getStoredUsers();
        if (users.find(u => u.username === username)) {
            return "ERROR: Username already exists";
        }
        if (users.find(u => u.email === email)) {
            return "ERROR: Email already registered";
        }

        const newUser = {
            id: users.length + 1,
            username,
            password: this.simulateHash(password),
            email
        };
        users.push(newUser);
        this.setStoredUsers(users);

        // Set current user
        localStorage.setItem('currentUser', JSON.stringify(newUser));

        return "SUCCESS: User registered successfully";
    }

    async Login(username, password) {
        await this.simulateDelay();

        const users = this.getStoredUsers();
        const user = users.find(u => u.username === username);

        if (!user) {
            return "ERROR: Invalid username or password";
        }

        if (this.simulateVerify(password, user.password)) {
            localStorage.setItem('currentUser', JSON.stringify(user));
            return "SUCCESS: Login successful";
        } else {
            return "ERROR: Invalid username or password";
        }
    }

    async Logout() {
        await this.simulateDelay();
        localStorage.removeItem('currentUser');
    }

    async AddEmployee(name, email, department, hireDate) {
        await this.simulateDelay();

        if (!this.isUserLoggedIn()) {
            return "ERROR: User not logged in";
        }

        const employees = this.getStoredEmployees();
        if (employees.find(e => e.email === email)) {
            return "ERROR: Employee with this email already exists";
        }

        const newEmployee = {
            id: employees.length + 1,
            name,
            email,
            department,
            hireDate: hireDate.toISOString(),
            absentDays: 0
        };

        employees.push(newEmployee);
        this.setStoredEmployees(employees);

        return "SUCCESS: Employee added successfully";
    }

    async GetAllEmployees() {
        await this.simulateDelay();

        if (!this.isUserLoggedIn()) {
            return [];
        }

        const employees = this.getStoredEmployees();
        // Calculate working days for each employee
        return employees.map(emp => ({
            ...emp,
            CalculatedWorkingDays: this.calculateWorkingDays(emp.hireDate, emp.absentDays)
        }));
    }

    async UpdateAbsentDays(employeeId, newAbsentDays) {
        await this.simulateDelay();

        if (!this.isUserLoggedIn()) {
            return "ERROR: User not logged in";
        }

        if (newAbsentDays < 0) {
            return "ERROR: Absent days cannot be negative";
        }

        const employees = this.getStoredEmployees();
        const employee = employees.find(e => e.id === employeeId);

        if (!employee) {
            return "ERROR: Employee not found";
        }

        const totalDaysSinceHire = this.calculateTotalDaysSinceHire(employee.hireDate);
        if (newAbsentDays > totalDaysSinceHire) {
            return `ERROR: Absent days (${newAbsentDays}) cannot exceed total days since hire (${totalDaysSinceHire})`;
        }

        employee.absentDays = newAbsentDays;
        this.setStoredEmployees(employees);

        return "SUCCESS: Absent days updated successfully";
    }

    async ExportEmployeesToCsv() {
        await this.simulateDelay();

        if (!this.isUserLoggedIn()) {
            return "ERROR: User not logged in";
        }

        const employees = await this.GetAllEmployees();
        const headers = "Id,Name,Email,Department,HireDate,AbsentDays,WorkingDays";
        const rows = employees.map(emp =>
            `${emp.id},${emp.name},${emp.email},${emp.department},${new Date(emp.hireDate).toISOString().split('T')[0]},${emp.absentDays},${emp.CalculatedWorkingDays}`
        );

        const csvContent = [headers, ...rows].join('\n');
        return `SUCCESS:${csvContent}`;
    }

    // Helper methods for simulated storage
    getStoredUsers() {
        const stored = localStorage.getItem('users');
        if (stored) {
            return JSON.parse(stored);
        }
        // Default admin user
        const defaultUsers = [{
            id: 1,
            username: 'admin',
            email: 'admin@company.com',
            password: this.simulateHash('admin123')
        }];
        this.setStoredUsers(defaultUsers);
        return defaultUsers;
    }

    setStoredUsers(users) {
        localStorage.setItem('users', JSON.stringify(users));
    }

    getStoredEmployees() {
        const stored = localStorage.getItem('employees');
        return stored ? JSON.parse(stored) : [];
    }

    setStoredEmployees(employees) {
        localStorage.setItem('employees', JSON.stringify(employees));
    }

    isUserLoggedIn() {
        return localStorage.getItem('currentUser') !== null;
    }

    // Simulation of BCrypt-like functionality
    simulateHash(password) {
        // In a real app, this would be proper BCrypt hashing on the server
        return btoa(password + '_hashed_simulated');
    }

    simulateVerify(password, hash) {
        return this.simulateHash(password) === hash;
    }

    calculateWorkingDays(hireDate, absentDays) {
        const totalDays = this.calculateTotalDaysSinceHire(hireDate);
        return Math.max(0, totalDays - absentDays);
    }

    calculateTotalDaysSinceHire(hireDate) {
        const hire = new Date(hireDate);
        const today = new Date();
        const diffTime = today - hire;
        return Math.floor(diffTime / (1000 * 60 * 60 * 24));
    }

    simulateDelay() {
        return new Promise(resolve => setTimeout(resolve, 500));
    }
}

File: EmployeeManagement_System/Services/DatabaseService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using EmployeeManagementSystem.Data;
using EmployeeManagementSystem.Models;
using EmployeeManagementSystem.Services;

namespace EmployeeManagementSystem.Services
{
    public class DatabaseService
    {
        private readonly AppDbContext _context;
        private int? _currentUserId;

        public DatabaseService(AppDbContext context)
        {
            _context = context;
            _context.Database.EnsureCreated();
        }

        public string Signup(string username, string password, string email)
        {
            try
            {
                // Check if user already exists
                if (_context.Users.Any(u => u.Username == username))
                {
                    return "ERROR: Username already exists";
                }

                if (_context.Users.Any(u => u.Email == email))
                {
                    return "ERROR: Email already registered";
                }

                // Hash password
                string hashedPassword = BCryptSimulatedHasher.HashPassword(password);

                // Create new user
                var newUser = new User
                {
                    Username = username,
                    PasswordHash = hashedPassword,
                    Email = email
                };

                _context.Users.Add(newUser);
                _context.SaveChanges();

                _currentUserId = newUser.Id;
                return "SUCCESS: User registered successfully";
            }
            catch (Exception ex)
            {
                return $"ERROR: Registration failed - {ex.Message}";
            }
        }

        public string Login(string username, string password)
        {
            try
            {
                var user = _context.Users.FirstOrDefault(u => u.Username == username);
                if (user == null)
                {
                    return "ERROR: Invalid username or password";
                }

                if (BCryptSimulatedHasher.VerifyPassword(password, user.PasswordHash))
                {
                    _currentUserId = user.Id;
                    return "SUCCESS: Login successful";
                }
                else
                {
                    return "ERROR: Invalid username or password";
                }
            }
            catch (Exception ex)
            {
                return $"ERROR: Login failed - {ex.Message}";
            }
        }

        public void Logout()
        {
            _currentUserId = null;
        }

        public bool IsUserLoggedIn()
        {
            return _currentUserId.HasValue;
        }

        public string AddEmployee(string name, string email, string department, DateTime hireDate, decimal monthlySalary, int workingDaysPerMonth = 22)
        {
            try
            {
                if (!IsUserLoggedIn())
                {
                    return "ERROR: User not logged in";
                }

                // Check if email already exists
                if (_context.Employees.Any(e => e.Email == email))
                {
                    return "ERROR: Employee with this email already exists";
                }

                var employee = new Employee
                {
                    Name = name,
                    Email = email,
                    Department = department,
                    HireDate = hireDate,
                    AbsentDays = 0,
                    MonthlySalary = monthlySalary,
                    WorkingDaysPerMonth = workingDaysPerMonth
                };

                _context.Employees.Add(employee);
                _context.SaveChanges();

                return "SUCCESS: Employee added successfully";
            }
            catch (Exception ex)
            {
                return $"ERROR: Failed to add employee - {ex.Message}";
            }
        }

        public List<Employee> GetAllEmployees()
        {
            if (!IsUserLoggedIn())
            {
                return new List<Employee>();
            }

            return _context.Employees.ToList();
        }

        public Employee GetEmployeeById(int id)
        {
            try
            {
                if (!IsUserLoggedIn())
                {
                    return null;
                }

                return _context.Employees.Find(id);
            }
            catch (Exception ex)
            {
                // Log error
                return null;
            }
        }

        public string UpdateEmployee(int id, string name, string email, string department, DateTime hireDate, decimal monthlySalary, int workingDaysPerMonth)
        {
            try
            {
                if (!IsUserLoggedIn())
                {
                    return "ERROR: User not logged in";
                }

                var employee = _context.Employees.Find(id);
                if (employee == null)
                {
                    return "ERROR: Employee not found";
                }

                // Check if email already exists for other employees
                if (_context.Employees.Any(e => e.Email == email && e.Id != id))
                {
                    return "ERROR: Another employee with this email already exists";
                }

                employee.Name = name;
                employee.Email = email;
                employee.Department = department;
                employee.HireDate = hireDate;
                employee.MonthlySalary = monthlySalary;
                employee.WorkingDaysPerMonth = workingDaysPerMonth;

                _context.SaveChanges();
                return "SUCCESS: Employee updated successfully";
            }
            catch (Exception ex)
            {
                return $"ERROR: Failed to update employee - {ex.Message}";
            }
        }

        public string DeleteEmployee(int id)
        {
            try
            {
                if (!IsUserLoggedIn())
                {
                    return "ERROR: User not logged in";
                }

                var employee = _context.Employees.Find(id);
                if (employee == null)
                {
                    return "ERROR: Employee not found";
                }

                _context.Employees.Remove(employee);
                _context.SaveChanges();
                return "SUCCESS: Employee deleted successfully";
            }
            catch (Exception ex)
            {
                return $"ERROR: Failed to delete employee - {ex.Message}";
            }
        }

        public string UpdateAbsentDays(int employeeId, int newAbsentDays)
        {
            try
            {
                if (!IsUserLoggedIn())
                {
                    return "ERROR: User not logged in";
                }

                if (newAbsentDays < 0)
                {
                    return "ERROR: Absent days cannot be negative";
                }

                var employee = _context.Employees.Find(employeeId);
                if (employee == null)
                {
                    return "ERROR: Employee not found";
                }

                var totalDaysSinceHire = (DateTime.Now - employee.HireDate).Days;
                if (newAbsentDays > totalDaysSinceHire)
                {
                    return $"ERROR: Absent days ({newAbsentDays}) cannot exceed total days since hire ({totalDaysSinceHire})";
                }

                employee.AbsentDays = newAbsentDays;
                _context.SaveChanges();

                return "SUCCESS: Absent days updated successfully";
            }
            catch (Exception ex)
            {
                return $"ERROR: Failed to update absent days - {ex.Message}";
            }
        }

        public string ExportEmployeesToCsv()
        {
            try
            {
                if (!IsUserLoggedIn())
                {
                    return "ERROR: User not logged in";
                }

                var employees = GetAllEmployees();
                var csvContent = new List<string>
                {
                    "Id,Name,Email,Department,HireDate,AbsentDays,WorkingDays,MonthlySalary,CalculatedSalary"
                };

                foreach (var emp in employees)
                {
                    csvContent.Add(
                        $"{emp.Id},{emp.Name},{emp.Email},{emp.Department},{emp.HireDate:yyyy-MM-dd},{emp.AbsentDays},{emp.CalculatedWorkingDays},{emp.MonthlySalary},{emp.CalculatedSalary}"
                    );
                }

                string csvData = string.Join(Environment.NewLine, csvContent);
                return $"SUCCESS:{csvData}";
            }
            catch (Exception ex)
            {
                return $"ERROR: Export failed - {ex.Message}";
            }
        }

        // Calculate salary for specific employee
        public decimal CalculateSalary(int employeeId, int absentDays)
        {
            var employee = GetEmployeeById(employeeId);
            if (employee == null || employee.WorkingDaysPerMonth == 0)
                return 0;

            var dailyRate = employee.MonthlySalary / employee.WorkingDaysPerMonth;
            var workingDays = employee.WorkingDaysPerMonth - absentDays;
            return dailyRate * Math.Max(0, workingDays);
        }
    }
}